File: android/src/main/kotlin/com/superwall/superwallkit_flutter/json/TriggerResultMapper.kt
import com.superwall.sdk.models.triggers.Experiment
import com.superwall.sdk.models.triggers.TriggerResult

fun TriggerResult.pigeonify(): PTriggerResult {
    return when (this) {
        is TriggerResult.PlacementNotFound -> PPlacementNotFoundTriggerResult()
        is TriggerResult.NoAudienceMatch -> PNoAudienceMatchTriggerResult()
        is TriggerResult.Paywall -> PPaywallTriggerResult(
            experiment = experiment.pigeonify()
        )
        is TriggerResult.Holdout -> PHoldoutTriggerResult(
            experiment = experiment.pigeonify()
        )
        is TriggerResult.Error -> PErrorTriggerResult(
            error = error.localizedMessage ?: "Unknown error"
        )
    }
}

fun Experiment.pigeonify(): PExperiment {
    return PExperiment(
        id = id,
        groupId = groupId,
        variant = variant.pigeonify()
    )
}

fun Experiment.Variant.pigeonify(): PVariant {
    return PVariant(
        id = id,
        type = when(this.type) {
            Experiment.Variant.VariantType.TREATMENT -> PVariantType.TREATMENT
            Experiment.Variant.VariantType.HOLDOUT -> PVariantType.HOLDOUT
        },
        paywallId = paywallId
    )
}

File: ios/superwallkit_flutter/Sources/superwallkit_flutter/Mappers/TriggerResultMapper.swift
import Foundation
import SuperwallKit

extension TriggerResult {
  func pigeonify() -> PTriggerResult {
    switch self {
    case .placementNotFound:
      return PPlacementNotFoundTriggerResult()
    case .noAudienceMatch:
      return PNoAudienceMatchTriggerResult()
    case .paywall(let experiment):
      return PPaywallTriggerResult(experiment: experiment.pigeonify())
    case .holdout(let experiment):
      return PHoldoutTriggerResult(experiment: experiment.pigeonify())
    case .error(let nSError):
      return PErrorTriggerResult(error: nSError.localizedDescription)
    }
  }
}

extension Experiment {
  func pigeonify() -> PExperiment {
    return PExperiment(
      id: id,
      groupId: groupId,
      variant: variant.pigeonify()
    )
  }
}

extension Experiment.Variant {
  func pigeonify() -> PVariant {
    return PVariant(id: id, type: type.pigeonify())
  }
}

extension Experiment.Variant.VariantType {
  func pigeonify() -> PVariantType {
    switch self {
    case .holdout:
      return .holdout
    case .treatment:
      return .treatment
    }
  }
}
